### PR TITLE
Add reference higher-order linear attention

### PR DESCRIPTION
**Pull request by Fr0do**
## Summary
- add a reference recurrent second-order Higher-order Linear Attention op
- expose HigherOrderLinearAttention as an FLA layer
- add tests comparing the streaming identity against the explicit masked quadratic form, cache chunking, backward gradients, and layer shape

## Notes
This is intentionally a PyTorch reference implementation first, matching the HLA masked streaming identity with state (S, C, m, G, h). It gives FLA a correct API/test target before adding a chunk/Triton/TileLang kernel.

## Validation
- python -m pytest tests/ops/test_hla.py -q

### CODE REVIEW

**Comment by Fr0do** [comment]
Addressed the Gemini/Codex review feedback in `a3e8061a`:

- `attention_mask` now masks `q`, `k`, and `v`, so padded tokens do not update `S`, `G`, or `h`.
- `HigherOrderLinearAttention` now explicitly supports only `mode="recurrent"` until a fused/chunk kernel is added.
- Removed `@torch.compiler.disable`; smoke-tested `torch.compile(recurrent_hla, mode="reduce-overhead")` on the H200 env.
- The normalized path now uses a signed epsilon floor for `den`, so negative denominators keep their sign instead of being clamped to `+eps`.

Paper/reference context: this PR implements the second-order strict-causal HLA recurrence from **Higher-order Linear Attention**, arXiv:2510.27258 (v3, 2026-05-14), plus the public project notes at https://yifanzhang-pro.github.io/HLA/.

For one head, the reference op maintains the prefix summaries:

```text
S_t = sum_{i<=t} k_i k_i^T
C_t = sum_{i<=t} q_i v_i^T
m_t = sum_{i<=t} q_i
G_t = sum_{i<=t} k_i k_i^T C_{i-1}
h_t = sum_{i<=t} k_i k_i^T m_{i-1}
```

The implemented output identities are:

```text
o_t = q_t^T (S_t C_t - G_t)
den_t = q_t^T (S_t m_t - h_t)
```

with the optional normalized variant using `o_t / signed_clamp_min(den_t, eps)`. This is equivalent to the explicit masked quadratic form covered by `tests/ops/test_hla.py`:

```text
W = tril(Q K^T)
A = tril(W W^T)
O = A V
```

but computed by the streaming state above, without materializing the `T x T` matrices.

Comparison to recent FLA PRs:

- Similar to #900 (Raven), this is a paper-linked new layer/operator path with package exports and focused tests. The difference is that HLA is intentionally reference-first and does not yet add a full model family/config.
- Compared with #905, this PR does not change shared op signatures or existing state-layout behavior; it is isolated under `fla.ops.hla` / `fla.layers.hla`.
- Compared with #916, this is not a Triton performance PR yet. The validation is correctness-focused now (`pytest tests/ops/test_hla.py -q`, 6 passed on H200), with a clean path for a follow-up chunk/Triton kernel and benchmark table once the recurrence/API is accepted.

**Comment by Fr0do** [comment]
Added a Triton forward kernel path in `204dfda1`.

Scope:
- New op: `fla.ops.hla.triton_recurrent_hla`.
- Registered in `benchmarks/ops` next to `recurrent_hla`.
- Tested against the PyTorch recurrent HLA reference for normalized and unnormalized outputs.
- It is intentionally **forward-only** and **not exposed as a `HigherOrderLinearAttention` layer mode** yet, because it does not implement autograd backward or recurrent cache state output. The trainable/cache-complete layer remains `mode="recurrent"`.

Validation on the H200 env:

```text
python -m pytest tests/ops/test_hla.py -q
8 passed
```

Direct H200 forward microbenchmarks, bf16, `CUDA_VISIBLE_DEVICES=2`.
The server was shared with other GPU workloads, so treat these as smoke-scale relative numbers rather than a final benchmark table.

```text
shape             op                         median ms
B1 T512 H4 D16    torch SDPA FlashAttention    0.038
B1 T512 H4 D16    chunk_retention              0.192
B1 T512 H4 D16    chunk_gla                    0.216
B1 T512 H4 D16    triton_recurrent_hla         0.427
B1 T512 H4 D16    chunk_gdn                    0.452
B1 T512 H4 D16    chunk_kda                    0.501
B1 T512 H4 D16    recurrent_hla               93.092

B1 T1024 H4 D16   torch SDPA FlashAttention    0.018
B1 T1024 H4 D16   chunk_retention              0.197
B1 T1024 H4 D16   chunk_gla                    0.209
B1 T1024 H4 D16   chunk_gdn                    0.387
B1 T1024 H4 D16   chunk_kda                    0.472
B1 T1024 H4 D16   triton_recurrent_hla         0.859
B1 T1024 H4 D16   recurrent_hla              195.159
```

FlashAttention note: the installed `flash_attn` package on this H200 env does not expose `flash_attn_func` (`flash_attn` is only a namespace with `cute`), so I used PyTorch SDPA forced to `SDPBackend.FLASH_ATTENTION` for the FA baseline. The standard benchmark registry still has the `flash_attn` entry for environments where the official callable is available.

Interpretation:
- The Triton kernel is about 215-227x faster than the PyTorch recurrent HLA reference on these shapes.
- It is still slower than highly optimized existing FLA chunk kernels for first-order linear attention/retention, which is expected: this is a single-program recurrent second-order HLA kernel, not a chunked parallel HLA algorithm.
- This gives us a concrete forward/inference baseline and a benchmark registration point for a follow-up autograd/chunked kernel.

**Comment by Fr0do** [comment]
Note: I amended the Triton HLA commit after the initial benchmark comment to remove the lint-only unused `offs_d` local. The current pushed commit is `ceb77fc0`.

**Comment by Fr0do** [comment]
I moved the Triton HLA kernel out of this upstream PR branch for now. The draft PR is back to the safer reference-HLA state at `a3e8061a`; Triton forward-kernel work is preserved separately on the fork-only branch `Fr0do:hla-triton-wip`. We will validate/backfill backward/cache support, or keep it benchmark-only, before reintroducing it upstream.

**Comment by Fr0do** [comment]
Closing this upstream PR for now. We are moving HLA reference/Triton iteration back to the fork for validation before re-opening anything against fla-org.
